detox/src/devices/Device.js

```diff
@@ -5,119 +5,95 @@ const { traceCall } = require('../utils/trace');
-    } else if (params.userActivity) {
-      this._createPayloadFileAndUpdatesParamsObject('userActivity', 'detoxUserActivityDataURL', params, baseLaunchArgs);
-    }
+  async prepare() {
+    await this.deviceDriver.prepare();
 
-    if (params.permissions) {
-      await this.deviceDriver.setPermissions(deviceId, bundleId, params.permissions);
-    }
+    this._deviceId = await traceCall('acquireDevice', () =>
+      this.deviceDriver.acquireFreeDevice(this._deviceConfig.device));
 
-    if (params.disableTouchIndicators) {
-      baseLaunchArgs['detoxDisableTouchIndicators'] = true;
+    const appAliases = Object.keys(this._appsConfig);
+    if (appAliases.length === 1) {
+      await this.selectApp(appAliases[0]);
     }
+  }
 
-    if (this._isAppInBackground(params, bundleId)) {
-      if (hasPayload) {
-        await this.deviceDriver.deliverPayload({...params, delayPayload: true});
-      }
+  async selectApp(name) {
+    if (name === undefined) {
+      throw this._errorBuilder.cantSelectEmptyApp();
     }
 
-    let processId;
-    if (this._behaviorConfig.launchApp === 'manual') {
-      processId = await this.deviceDriver.waitForAppLaunch(deviceId, bundleId, this._prepareLaunchArgs(baseLaunchArgs), params.languageAndLocale);
-    } else {
-      processId = await this.deviceDriver.launchApp(deviceId, bundleId, this._prepareLaunchArgs(baseLaunchArgs), params.languageAndLocale);
-      await this.deviceDriver.waitUntilReady();
-      await this.deviceDriver.waitForActive();
+    if (name === null) { // Internal use to unselect the app
+      this._currentApp = null;
+      return;
     }
-    this._processes[bundleId] = processId;
 
-    await this._emitter.emit('appReady', {
-      deviceId,
-      bundleId,
-      pid: processId,
-    });
-
-    if(params.detoxUserNotificationDataURL) {
-      await this.deviceDriver.cleanupRandomDirectory(params.detoxUserNotificationDataURL);
-    }
-
-    if(params.detoxUserActivityDataURL) {
-      await this.deviceDriver.cleanupRandomDirectory(params.detoxUserActivityDataURL);
+    const appConfig = this._appsConfig[name];
+    if (!appConfig) {
+      throw this._errorBuilder.cantFindApp(name);
     }
-  }
-
-  get appLaunchArgs() {
-    return this._launchArgs;
-  }
 
-  get id() {
-    return this._deviceId;
+    this._currentApp = appConfig;
+    this._cachedAppLaunchArgs = null;
+    await this._inferBundleIdFromBinary();
   }
 
-  get name() {
-    return this.deviceDriver.name;
+  async launchApp(params = {}, bundleId = this._bundleId) {
+    return traceCall('launchApp', () => this._doLaunchApp(params, bundleId));
   }
 
-  get type() {
-    return this._deviceConfig.type;
+  /**
+   * @deprecated
+   */
+  async relaunchApp(params = {}, bundleId) {
+    if (params.newInstance === undefined) {
+      params['newInstance'] = true;
+    }
+    await this.launchApp(params, bundleId);
```

Comment:
Sign this deprecation up for Detox 19 😄

Reply:
AFAIR, we spoke with Leo about this — and looks like we might refrain from deprecating it at all.